tests/test_TFNetworkLayer.py

```diff
@@ -868,6 +868,54 @@ def test_pool_layer_NCHW():
+    net = TFNetwork(extern_data=ExternData())
+    net.train_flag = True
+    with tf.variable_scope("src_nchw"):
+      src_nhwc = InternalLayer(name="src_nchw", network=net, out_type={"dim": 16,
+                                                                       "shape": (16, None, 16),
+                                                                       "batch_dim_axis": 0,
+                                                                       "time_dim_axis": 2,
+                                                                       "feature_dim_axis": 1,
+                                                                       "sparse": False
+                                                                       })
+      src_nhwc.output.placeholder = tf.placeholder(shape=(None, 16, None, 16), dtype=tf.float32)
+      src_nhwc.output.size_placeholder = {1: tf.placeholder(shape=(None,), dtype=tf.int32)}
+
+    with tf.variable_scope("batch_norm_masked_nchw"):
+      batch_norm_1 = BatchNormLayer(name="batch_norm_masked_nchw", network=net, masked_time=False,
+                                    sources=[src_nhwc],
+                                    output=BatchNormLayer.get_out_data_from_opts(name="batch_norm_masked_nchw",
+                                                                                 sources=[src_nhwc],
+                                                                                 network=net))
+      batch_norm_1.post_init(layer_desc=None)
+
+    with tf.variable_scope("batch_norm_nonmasked_nchw"):
+      batch_norm_2 = BatchNormLayer(name="batch_norm_nonmasked_nchw", network=net, masked_time=False,
+                                    sources=[src_nhwc],
+                                    output=BatchNormLayer.get_out_data_from_opts(name="batch_norm_nonmasked_nchw",
+                                                                                 sources=[src_nhwc],
+                                                                                 network=net))
+      batch_norm_2.post_init(layer_desc=None)
+
+    tf.global_variables_initializer().run()
+    out, seq_lens = session.run([batch_norm_1.output.placeholder,
+                                 batch_norm_1.output.size_placeholder[1]],
+                                feed_dict={src_nhwc.output.placeholder: np.random.rand(10, 16, 11, 16),
+                                           src_nhwc.output.size_placeholder[1]: np.full(shape=(10,), fill_value=11)}
+                                )
+    print(out.shape)
+    out, seq_lens = session.run([batch_norm_2.output.placeholder,
+                                 batch_norm_2.output.size_placeholder[1]],
+                                feed_dict={src_nhwc.output.placeholder: np.random.rand(10, 16, 11, 16),
+                                           src_nhwc.output.size_placeholder[1]: np.full(shape=(10,), fill_value=11)}
+                                )
+    print(out.shape)
```

Comment:
The test is kind of incomplete.
You should also check that the output is correct.
You should also prepare the input in such a way that the output should be the same (of `batch_norm_1` and `batch_norm_2`), and then do an actually assert on that.
Also, the test should cover a more interesting case where the seq lens are not all full, but e.g. some of them are also less than full (e.g. 5 or so). And then prepare the input data in such a way that it's zero for the other values, so that actually `masked_time=True` and `masked_time=False` really gives the same result.

Reply:
It's not clear how to do this. If the seqs are not full, the result will be different. And filling the data with 0s wouldn't solve this (mean won't change but variance will).

Reply:
I think, if seq lens are not full, masking has to be applied

Reply:
Yes, but that's the point. Only in that case, the test make sense. And yes, we actually discussed this together, of course the result is different then, but we discussed how to get from one number to the other. Instead of filling with 0s, fill it such that it will become the same.